p2panda-net/src/sync/mod.rs

```
@@ -141,59 +155,87 @@ where
+                        .send(ToEngineActor::SyncFailed {
+                            peer,
+                            topic: topic.clone(),
+                        })
+                        .await
+                        .expect("engine channel closed");
+                },
+                Some(message) = rx.recv() => {
+                    // 1. Handshake Phase.
+                    // ~~~~~~~~~~~~~~~~~~~
+                    //
+                    // At the beginning of every sync session the "accepting" peer needs to learn the
+                    // topic of the "initiating" peer during the handshake phase. This is the first
+                    // message we're expecting:
+                    if let FromSync::HandshakeSuccess(handshake_topic) = &message {
+                        // It should only be sent once so topic should be `None` now.
+                        if topic.is_some() {
+                            error!("topic message already received");
+                            break;
+                        }
+
+                        topic = Some(handshake_topic.clone());
+
+                        // Inform the engine that we are expecting sync messages from the peer on
+                        // this topic.
+                        engine_actor_tx
+                            .send(ToEngineActor::SyncHandshakeSuccess {
+                                peer,
+                                topic: handshake_topic.clone(),
+                            })
+                            .await
+                            .expect("engine channel closed");
+
+                        continue;
+                    }
+
+                    // 2. Data Sync Phase.
+                    // ~~~~~~~~~~~~~~~~~~~
+                    //
+                    // The topic must be known at this point in order to process further messages.
+                    let Some(topic) = &topic else {
+                        error!("topic not received");
```

This silently fails when the handshake phase was not fullfilled, I assume we should catch that as well as this indicates that the other peer didn't follow the protocol?

---

It's definitely a protocol violation. The question for me is: what do we do with that information? From the perspective of the engine, no sync session has actually started in this case; it only "starts" when the accepting peer receives the topic and then sends the `ToEngineActor::SyncHandshakeSuccess` event.

I can imagine potentially having a "naughty list" for peers who repeatedly try to run sync sessions with faulty protocols. In that case, we might want to be reporting this peer to some higher-level (the sync connection handler maybe, or the engine). But we would still only know the peer (not the topic).

---

> It's definitely a protocol violation. The question for me is: what do we do with that information?

I'd think that it's more about "how can I avoid future me to deal with bugs and footguns" and a clear API surface with errors is one step towards that, logging usually only helps if you know what you're looking for.

I don't want to suggest this needs urgent fixing but lets maybe do a further ticket for it?